craft_parts/plugins/python_v2/python_plugin.py

```diff
@@ -0,0 +1,125 @@
+
+
+class PythonPluginProperties(PluginProperties, frozen=True):
+    """The part properties used by the python plugin."""
+
+    plugin: Literal["python"] = "python"
+
+    python_requirements: list[str] = []
+    python_packages: list[str] = []
+
+    # part properties required by the plugin
+    source: str  # pyright: ignore[reportGeneralTypeIssues]
+
+
+class PythonPlugin(Plugin):
+    """Python plugin, version 2.
+
+    Unlike the original Python plugin, this one does *not* require the creation of
+    virtual environments to work. Instead, it uses pip's support for both a) installing
+    directly into the user's base dir and b) using a Python interpreter that is not the
+    one running pip initially. This allows lighter payloads as the -venv package is
+    not required, and interacts well with usrmerged directories.
+
+    The downside is that the payload generated by the plugin assumes that the primed
+    contents will include the Python interpreter; if this is not the case, the runtime
+    will need to add the payload to Python's path (e.g. via PYTHONPATH).
+    """
+
+    properties_class = PythonPluginProperties
+    _options: PythonPluginProperties
+
+    @override
+    def get_build_snaps(self) -> set[str]:
+        """Return a set of required snaps to install in the build environment."""
+        return set()
+
+    @override
+    def get_build_packages(self) -> set[str]:
+        """Return a set of required packages to install in the build environment."""
+        return set()
+
+    @override
+    def get_build_environment(self) -> dict[str, str]:
+        """Return a dictionary with the environment to use in the build step."""
+        return {
+            "PIP_USER": "1",
+            "PYTHONUSERBASE": str(self._part_info.part_install_dir),
+            "PIP_BREAK_SYSTEM_PACKAGES": "1",
+            "PIP_PYTHON": "$(which python3)",
+        }
+
+    @override
+    def get_build_commands(self) -> list[str]:
+        """Return a list of commands to run during the build step."""
+        pip_calls: list[str] = []
+
+        # Disallow using the system's Python interpreter for now. The issue is that
+        # using the system interpreter will make pip skip dependencies that are already
+        # installed in the system, even though they won't be part of the part's files.
+        python_check = dedent("""\
+          if [[ ${PIP_PYTHON} == /usr/bin/* ]]; then
+            echo "Using the system Python interpreter is not supported." 1>&2
+            exit 1
+          fi
+        """)
+
+        # First the requirements
+        pip_calls.extend(
+            f"pip install -r {req}" for req in self._options.python_requirements
+        )
+
+        # Then any extra packages
+        if self._options.python_packages:
+            python_packages = " ".join(
+                [shlex.quote(pkg) for pkg in self._options.python_packages]
+            )
+            pip_calls.append(f"pip install {python_packages}")
```

Comment:
I'd rather do all three of these installs in one pip command so it errors if we have conflicts rather than letting the last command override.